### ISSUE TITLE
limiter plugin http error Too Many Requests and debug log shows searx.plugins.limiter: got connection=close when using via uwsgi Unix Domain Socket

### ISSUE THREAD

**Issue by ravermeister** [searxng/searxng#1289]
**Version of SearXNG, commit number if you are using on master branch and stipulate if you forked SearXNG**
```
git log -1
commit ea0cddba0b9366326eaebc0a3a55502a17e5baf4 (HEAD -> master, origin/master, origin/HEAD)
Merge: d633c1c6 1a8ff10e
Author: Alexandre Flament <alex@al-f.net>
Date:   Mon Jun 6 10:59:11 2022 +0200

    Merge pull request #1047 from return42/redis-lib
    
    Add a redis library to generalize DB functions we need in SearXNG.
``` 

**How did you install SearXNG?**
I installed via the official [Step by Step](https://docs.searxng.org/admin/installation-searxng.html) Guide

**What happened?**
When I enabled the limiter Plugin I can't execute any search. neither with [Brave](https://brave.com/) nor with [Waterfox](https://www.waterfox.net/)

It is not possible to use the limiter Plugin with these Browsers, as any search results in a HTTP Error 429 and in the log you can find this: `DEBUG   searx.plugins.limiter         : got Connection=close`

**How To Reproduce**

1. enable redis
```
redis:
  url: unix:///path/to/redis/sock?db=1

```
2. enable limiter plugin 
```
server:
...
  # monitor rate limits and protect privacy
  limiter: true
...
```
**Expected behavior**
Searching is possible with the above Browsers

**Screenshots & Logs**

**Additional context**

dalf found the line where the Logging entry is produced: [searx/plugins/limiter.py#65](https://github.com/searxng/searxng/blob/ea0cddba0b9366326eaebc0a3a55502a17e5baf4/searx/plugins/limiter.py#L65)

The Response Header are as follows:
```
content-length: 0
content-type: text/html; charset=utf-8
date: Mon, 06 Jun 2022 09:52:58 GMT
referrer-policy: strict-origin-when-cross-origin
server-timing: total;dur=3.412, render;dur=0
strict-transport-security: max-age=31536000
x-content-type-options: nosniff
x-download-options: noopen
x-frame-options: SAMEORIGIN
x-robots-tag: noindex, nofollow
x-xss-protection: 1; mode=block
```
The Request Header are as follows:
```
:authority: searx.url.domain
:method: GET
:path: /search?q=search_query
:scheme: https
accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
accept-encoding: gzip, deflate, br
accept-language: de-DE;q=0.7
cookie: categories=general; language=de-DE; locale=de; autocomplete=google; image_proxy=1; method=POST; safesearch=0; theme=simple; results_on_new_tab=0; doi_resolver=oadoi.org; simple_style=auto; query_in_title=; infinite_scroll=1; disabled_engines=; enabled_engines=; disabled_plugins=; enabled_plugins=; tokens=; maintab=on; enginetab=on
sec-fetch-dest: document
sec-fetch-mode: navigate
sec-fetch-site: none
sec-fetch-user: ?1
sec-gpc: 1
upgrade-insecure-requests: 1
user-agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/102.0.5005.61 Safari/537.36
```

**Comment by ononoki1** [comment]
Did you use HTTP/3?

**Comment by ravermeister** [comment]
I don't think so I'm using the `use_default_settings: true`. Looking in the base config there is only  `enable_http2: true` in the `outgoing` section set. But I've set  `http_protocol_version: "1.1"` in the server section. I'm using 
```
$ python --version
Python 3.9.13
```
and the morty proxy by the way

**Comment by ononoki1** [comment]
I mean inbound connection. Do you use HTTP/3 when sending request to searx.url.domain?

**Comment by return42** [comment]
> Did you use HTTP/3?

I think the issue is related to the limiter:

https://github.com/searxng/searxng/blob/ea0cddba0b9366326eaebc0a3a55502a17e5baf4/searx/plugins/limiter.py#L65-L67

This is at least, what the log indicates:

    DEBUG searx.plugins.limiter : got Connection=close

What me wonders; there is no `Connection=close` in the request headers shown in https://github.com/searxng/searxng/issues/1289#issue-1261614307

Does this `Connection=close` come from the reverse proxy?

----

Are you using nginx? .. is there a `proxy_set_header   Connection  ...` in your nginx site?

**Comment by ravermeister** [comment]
I can't tell exactly t.b.h I'm running this instance on [uberspace](https://uberspace.de/) via [Web Backend](https://blog.uberspace.de/web-backends-websockets/) but I can control header that should be added... Could it help to set a connection keepalive header explicitly? for both my search domain and morty domain?

I just stumbled upon: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Keep-Alive

So could it that Keep-alive is not allowed when http2 is enabled? (it is enabled via the default config)

**Comment by ononoki1** [comment]
I don't know what exactly you can do since I'm not familiar with uberspace, but you can first try to add `Connection: ""` header for reverse proxy.

**Comment by return42** [comment]
> So could it that Keep-alive is not allowed when http2 is enabled? (it is enabled via the default config)

Yes, this header should not be set .. and it is suspicious for the limiter if set to `close`. 

Some WEB client (e.g. my FFox) do set this header, but the value is `keep-alive` (what is not suspicious to the limiter).

BTW: your link points to a `Keep-Alive` header / here we talking about the value `keep-alive` vs. `close` in the `Connection` header: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Connection

----

https://datatracker.ietf.org/doc/html/rfc7540#section-8.1.2.2

**Comment by ravermeister** [comment]
I tested all combinations.. (enable_http2 to true|false and setting Connection: keep-alive, and empty). But I see [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Connection#directives) that Connection `close` is the default for http 1.0 Could it be that the request is somehow "downgraded" allthough I've set http_protocol_version to 1.1 in the server config?

**Comment by return42** [comment]
"downgraded" IDK .. but I don't think so (if so, limiter would block all requests .. makes no sense).. lets resume: you have this issue with Brave and Waterfox browser but not when using FFox or Chrome?

**Comment by ravermeister** [comment]
with Firefox too. I don't have Chrome to test, but I would guess it happens on all Browsers.

**Comment by return42** [comment]
TBH I don't have a clue anymore, but one last question: What is the request header when you use FFox?

My FFox adds `Connection: keep-alive` ..

```
GET /searx/search?q=foo HTTP/1.1
Host: darmarit.org
User-Agent: Mozilla/5.0 (X11; Ubuntu; Linux x86_64; rv:100.0) Gecko/20100101 Firefox/100.0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8
Accept-Language: de-DE,en-US;q=0.7,en;q=0.3
Accept-Encoding: gzip, deflate, br
DNT: 1
Connection: keep-alive
Upgrade-Insecure-Requests: 1
Sec-Fetch-Dest: document
Sec-Fetch-Mode: navigate
Sec-Fetch-Site: none
Sec-Fetch-User: ?1
Pragma: no-cache
Cache-Control: no-cache
```

**Comment by ravermeister** [comment]
@return42 yes in firefox the `Connection: keep-alive` header is added, but the result is the same error. I'm out of Ideas too, 
would be cool to have a possibility to specify a 'blacklist' in the settings.yml for some limiter checks, so that one can use it even in such weird environments ;)

**Comment by return42** [comment]
> would be cool to have a possibility to specify a 'blacklist' in the settings.yml for some limiter checks

Yes, would be cool .. [I thought it might be better we add a configuration option to select what position in the `X-Forwarded-For` identifies the IP of the request (client).](https://github.com/searxng/searxng/issues/1237#issuecomment-1148340223) .. ore alternatively options: [Why not use `X-Real-IP` to replace `X-Forwarded-For`?](https://github.com/searxng/searxng/issues/1237#issuecomment-1147564516)

**Comment by sevmonster** [comment]
> I tested all combinations.. (enable_http2 to true|false and setting Connection: keep-alive, and empty). But I see [here](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Connection#directives) that Connection `close` is the default for http 1.0 Could it be that the request is somehow "downgraded" allthough I've set http_protocol_version to 1.1 in the server config?

The `Connection` header is entirely the responsibility of whatever proxy searxng is behind. It is a [hop-by-hop header](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.10) and as such should never be passed verbatim from the client—if it is being passed like that, the proxy responsible is going against spec and is broken. For that reason, there is nothing to be gained by attempting to troubleshoot the browser.

According to Uberspace, their Web Backend technology uses nginx under the hood with `proxy_pass`. If searxng is seeing `Connection: close`, nginx was not properly configured for keepalive connections. By default nginx's proxy module [uses HTTP/1.0](https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_http_version) and [explicitly disables HTTP persistent connections with `Connection: close`](https://nginx.org/en/docs/http/ngx_http_proxy_module.html#proxy_set_header). Since nginx's proxy module [only supports HTTP/1.1 and below](https://trac.nginx.org/nginx/ticket/923), your best and most performant option here is just to [enable upstream keepalive, set proxy protocol to HTTP/1.1, and remove the `Connection` header](https://nginx.org/en/docs/http/ngx_http_upstream_module.html#keepalive). At the very least, remove the `Connection` header; in nginx this is done by setting it to the empty string i.e. `proxy_set_header Connection "";`.

If you do not have direct access to your nginx configuration, you need to read their documentation or contact their support to see how to update it.

**Comment by return42** [comment]
@ravermeister does your issue resist or can we close this issue?

**Comment by ravermeister** [comment]
the problem still persists, but according to the last post it is because of the nginx Configuration of my hoster. 
I will contact their support and will re-open if required  :smiley: 

So yes, you can close it if you want

**Comment by ravermeister** [comment]
after I installed my own nginx instance and configured the proxy as follows (uwsgi is listening on 8010):
`
        location / {
            proxy_set_header Host searx.rimkus.it;
            proxy_pass http://localhost:8010;
            proxy_http_version 1.1;
            proxy_set_header Connection "";
        }

`
it works as expected. Next challenge: try to proxy via socket file :)

**Comment by sevmonster** [comment]
Don't forget the other headers, the limiter plugin will not be able to keep track of connection sources without them.

https://docs.searxng.org/admin/installation-nginx.html

**Comment by ravermeister** [comment]
@sevmonster Thanks. it is working when I use searxNG via uwsgi http. But if I use it via Unix Socket file, the error occurs again. 
I already tried adding 
```
so-keepalive=true
http-keepalive=1
add-header=Connection: Keep-Alive
```
to the uwsgi config of searxNG without success. Do you (or anoybody else) have an Idea, how I can get it running with Unix Domain Socket file?

**Comment by sevmonster** [comment]
You didn't share your latest configuration so it's not really possible to give any advice here besides shots in the dark.

And this issue seems potentially different enough that you may want to open another issue for it.

Note that, as I touched on earlier, `Connection: Keep-Alive` is not valid for HTTP/1.1 as keepalive is default and the `Keep-Alive` header is no longer in the spec. Don't include it if you are using HTTP/1.1. (Since you didn't share your config aside from partial uwsgi I don't know if you are.)

**Comment by ravermeister** [comment]
Closed in favor of #1492 (as the original error via uwsgi http is working)
